src/components/createContainer.jsx

```diff
@@ -32,41 +34,47 @@ export default function createContainer(DecoratedComponent, specs) {
     static DecoratedComponent = DecoratedComponent
 
     static contextTypes = {
-      renderLoading: PropTypes.func.isRequired,
-      adaptor: adaptorShape.isRequired,
+      adrenaline: PropTypes.shape({
+          renderLoading: PropTypes.func.isRequired,
+          adaptor: adaptorShape.isRequired
+      }).isRequired
+    }
+
+    static childContextTypes = {
+        adrenaline: containerShape.isRequired
     }
 
     constructor(props, context) {
       super(props, context);
-
       this.isNotMounted = true;
+      this.state = {};
+    }
 
-      const { adaptor } = context;
-      //DecoratedComponent.prototype.shouldComponentUpdate = adaptor.shouldComponentUpdate;
-
-      const initialArgs = !!specs.args ? specs.args(props) : {};
-
-      this.state = {
-        data: {},
-        args: initialArgs,
-      };
-
-      const dataKeys = Object.keys(specs.queries(initialArgs));
-      const { propTypes } = DecoratedComponent;
-      const requiredKeys = dataKeys.filter(key =>
-        propTypes[key] && !propTypes[key].isRequired
-      );
-
-      this.isDataLoaded = (data = {}) => requiredKeys.every(key => {
-        return data[key] !== null && data[key] !== undefined
-      });
+    getChildContext() {
+        const { adrenaline } = this.context;
+        const container = {
```

Comment:
I think it's better to utilize context for passing this stuff down to decorated components. This way we avoid prop conflicts and we don't pollute the decorated component prop updates with adrenaline's internal state changes. It should make it much easier to reason about prop changes inside the container components being decorated.